usaspending_api/search/filters/postgres/HierarchicalFilter.py

```diff
@@ -0,0 +1,86 @@
+from abc import abstractmethod
+from django.db.models import Q
+
+
+class HierarchicalFilter:
+    @classmethod
+    def _query_string(cls, queryset, require, exclude):
+        positive_nodes = [
+            cls.node(code, True, require, exclude) for code in require if cls._has_no_parents(code, require + exclude)
+        ]
+
+        negative_nodes = [
+            cls.node(code, False, require, exclude) for code in exclude if cls._has_no_parents(code, require + exclude)
+        ]
+
+        q = Q()
+        for node in positive_nodes:
+            q |= node.get_query()
+        for node in negative_nodes:
+            q &= node.get_query()
+
+        queryset = queryset.filter(q)
+        return queryset
+
+    @classmethod
+    def _has_no_parents(cls, code, other_codes):
+        return not len([match for match in other_codes if cls.code_is_parent_of(match, code)])
+
+    @staticmethod
+    def code_is_parent_of(code, other):
+        return other[: len(code)] == code and len(code) < len(other)
+
+    @staticmethod
+    @abstractmethod
+    def node(code, positive, positive_codes, negative_codes):
+        pass
+
+
+class Node:
+    """Represents one part of the final query, either requiring or excluding one code, with any exceptions"""
+
+    code: str
+    ancestors: list
+    positive: bool
+    children: list
+
+    def __init__(self, code, positive, positive_codes, negative_codes):
+        self.code = code[-1]
+        self.ancestors = code[:-1]
+        self.positive = positive
+        self.populate_children(positive_codes, negative_codes)
+
+    def populate_children(self, positive_codes, negative_codes):
+        self.children = []
+        self._pop_children_helper(positive_codes, True, positive_codes, negative_codes)
+        self._pop_children_helper(negative_codes, False, positive_codes, negative_codes)
+
+    def _pop_children_helper(self, codes, is_positive, positive_codes, negative_codes):
+        for other_code in codes:
+            if self.is_parent_of(other_code):
+                self.children.append(self.clone(other_code, is_positive, positive_codes, negative_codes))
+
+    def get_query(self) -> Q:
+        if self.positive:
+            filter = self._basic_search_unit()
+        else:
+            filter = ~self._basic_search_unit()
```

Comment:
@kbard is this code impacted by the Django ORM issue you raised earlier this week?

Reply:
Hmm.  That applied to negative filters in aggregations.  This is for advanced search, right?  I think everything's pre-aggregated in the matviews?  I _think_ it should be ok, but I'd have to look at the generated SQL to be sure.

Reply:
Ok.  Checked out the SQL.  This is definitely not affected by that issue.  This bit of code is just pulling the TAS ids.  No aggregations.

Having said that, though, it is very easy to make the server throw a 500 by submitting something like `"include": [["028", "028-0401", "028-2017/2017-04[1-000"]]`which causes malformed regex in the query.

@BrianZito

Reply:
Funny thing is one of the last things I did before putting the PR up was to remove the validation that blocks things like that since I assumed this was just an ES issue. I suppose the best solution is to just be stricter again.